### PR TITLE
Solaris: fixes build and tests, adds CI

### PR DESCRIPTION
**Pull request by psumbera**
Disables some tests for Solaris.

test/test_sendfile.rs:

Solaris, sendfilev() doesn't support AF_UNIX sockets. Instead, it expects an AF_INET or AF_INET6 sockets.

### CODE REVIEW

**Review by SteveLauC — COMMENTED** [review]
Thanks for the PR! Sorry for a bunch of comments 🫣

**Comment by SteveLauC** [comment]
BTW, [looks like `x86_64-sun-solaris` was renamed to `x86_64-pc-solaris`](https://github.com/rust-lang/rust/issues/85098), so this PR can close #935?

File: src/dir.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ use libc::{dirent, readdir_r};
 /// let mut cwd = Dir::open(".", OFlag::O_RDONLY | OFlag::O_CLOEXEC, Mode::empty()).unwrap();
 /// for res_entry in cwd.iter() {
 ///     let entry = res_entry.unwrap();
-///     println!("File name: {}", entry.file_name().to_str().unwrap());
```
[the remark this answers is not in the document]
Comment (psumbera) [reply src/dir.rs:49]: Yes, it panics: 
```
thread 'main' panicked at src/dir.rs:12:59:
called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 7, error_len: Some(1) }
```

File: src/sys/termios.rs [hunk 2]
```diff
@@ -463,6 +463,10 @@ impl SpecialCharacterIndices {
 pub use libc::NCCS;
 #[cfg(any(linux_android, target_os = "aix", bsd))]
 pub use libc::_POSIX_VDISABLE;
+// Solaris can use libc::_POSIX_VDISABLE once following is pulled in:
+// https://github.com/rust-lang/libc/pull/4103
+#[cfg(target_os = "solaris")]
+pub const _POSIX_VDISABLE: u8 = 0;
```
[the remark this answers is not in the document]
Comment (psumbera) [reply src/sys/termios.rs:469]: With git `libc` version I'm getting:
```
error[E0015]: cannot call non-const fn `CMSG_SPACE` in constant functions
   --> src/sys/socket/mod.rs:652:14
    |
652 |     unsafe { libc::CMSG_SPACE(mem::size_of::<T>() as libc::c_uint) as usize }
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants
```
No idea from where it comes...

[the remark this answers is not in the document]
Comment (SteveLauC) [reply src/sys/termios.rs:469]: Weird, looks like it is still a const function:

```sh
$ rg "CMSG_SPACE"
mod.rs
2508:    pub {const} fn CMSG_SPACE(length: ::c_uint) -> ::c_uint {
```

File: test/sys/test_timer.rs [hunk 3]
```diff
@@ -20,6 +20,9 @@ pub extern "C" fn handle_sigalarm(raw_signal: libc::c_int) {
 }
 
 #[test]
+// Solaris: The timer_create(3C) function requires the PRIV_PROC_CLOCK_HIGHRES
+// privilege to create a timer using the high-resolution system clock.
+#[cfg(not(target_os = "solaris"))]
```
[the remark this answers is not in the document]
Comment (psumbera) [reply test/sys/test_timer.rs:25]: Yes. It works. So I will remove it.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -327,6 +327,27 @@ jobs:
...
+          copyback: false
+          prepare: |
+            source <(curl -s https://raw.githubusercontent.com/psumbera/solaris-rust/refs/heads/main/sh.rust-web-install)
+            echo "~~~~ rustc --version ~~~~"
+            rustc --version
+            echo "~~~~ Solaris-version ~~~~"
+            uname -a
+          run: |
+            export PATH=$HOME/.rust_solaris/bin:$PATH
+            cargo build --target x86_64-pc-solaris --all-targets --all-features && cargo test
```
[the remark this answers is not in the document]
Comment (psumbera) [reply .github/workflows/ci.yml]: Seems that for Solaris the only test which requires `sudo` is `sys::test_timer::alarm_fires`.

I have added just `sudo` command for `cargo test`. I hope we don't need to run other targets for Solaris now.

[the remark this answers is not in the document]
Comment (SteveLauC) [reply .github/workflows/ci.yml]: Get it

File: src/sys/termios.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SteveLauC) [reply src/sys/termios.rs:469]: Could you please try this in our CI so that I can see the error message

[the remark this answers is not in the document]
Comment (SteveLauC) [reply src/sys/termios.rs]: Looks like libc PR https://github.com/rust-lang/libc/pull/4063 broke this, I verified that enabling feature `const-extern-fn` will work, but I think it is an issue, so let's wait for the upstream to fix it. 😶

```toml
libc = { git = "https://github.com/rust-lang/libc", branch = "libc-0.2", features = ["extra_traits", "const-extern-fn"] }
```

[the remark this answers is not in the document]
Comment (SteveLauC) [reply src/sys/termios.rs]: Upstream issue: https://github.com/rust-lang/libc/issues/4115

**Comment by SteveLauC** [comment]
> Hurd CI failure seems to be related to [rust-lang/libc#3785](https://github.com/rust-lang/libc/pull/3785).

Right, std is broken due to that PR

File: src/sys/socket/mod.rs [hunk 5]
```diff
@@ -1679,12 +1679,9 @@ impl ControlMessage<'_> {
 /// let localhost = SockaddrIn::from_str("1.2.3.4:8080").unwrap();
 /// let fd = socket(AddressFamily::Inet, SockType::Datagram, SockFlag::empty(),
 ///     None).unwrap();
-/// let (r, w) = pipe().unwrap();
 ///
 /// let iov = [IoSlice::new(b"hello")];
-/// let fds = [r.as_raw_fd()];
-/// let cmsg = ControlMessage::ScmRights(&fds);
-/// sendmsg(fd.as_raw_fd(), &iov, &[cmsg], MsgFlags::empty(), Some(&localhost)).unwrap();
+/// sendmsg(fd.as_raw_fd(), &iov, &[], MsgFlags::empty(), Some(&localhost)).unwrap();
```
[the remark this answers is not in the document]
Comment (SteveLauC) [reply src/sys/socket/mod.rs:1684]: So there is no control message we can use? 👀

Update: I just checked [enum ControlMessage](https://github.com/nix-rust/nix/blob/db4ce6aef3ee8e38724385377910b54efa1fddaa/src/sys/socket/mod.rs#L1168), looks like this `ScmRights` is available on Solaris, any reason to remove it?

**Comment by SteveLauC** [comment]
I just disabled the Hurd CI, you can rebase your branch to fix the issue:)

File: src/sys/socket/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SteveLauC) [reply src/sys/socket/mod.rs:1684]: @psumbera What about this one? 👀

**Review by SteveLauC — APPROVED** [review]
Thanks!

**Comment by SteveLauC** [comment]
> @SteveLauC any chance for new `nix` release? This commit is needed for Solaris Rust build.

I will give myself a push this weekend, but I cannot guarantee, I will try...

**Comment by psumbera** [comment]
> > @SteveLauC any chance for new `nix` release? This commit is needed for Solaris Rust build.
> 
> I will give myself a push this weekend, but I cannot guarantee, I will try...

Just gentle ping here. I really don't know how to make Rust language build to use latest NIx sources.

**Comment by SteveLauC** [comment]
0.30.0 released: https://crates.io/crates/nix/0.30.0

**Pull request merged**